### PR TITLE
OVModelForVisionCausalLM

### PR DESCRIPTION
**Pull request by eaidova**
# What does this PR do?

Enables conversion and inference for multimodality llm like llava, llava-next, falcon-vl, pixtral, internvl
Example of usage:

```
from PIL import Image
import requests
from optimum.intel.openvino import OVModelForVisualCausalLM
from transformers import AutoProcessor

model_id = "llava-hf/llava-v1.6-mistral-7b-hf"
model = OVModelForVisualCausalLM.from_pretrained(model_id)
image_file = "http://images.cocodataset.org/val2017/000000039769.jpg"
processor = AutoProcessor.from_pretrained(model_id)

conversation = [
    {
        "role": "user",
        "content": [
            {"type": "text", "text": "What are these?"},
            {"type": "image"},
        ],
    },
]
prompt = processor.apply_chat_template(conversation, add_generation_prompt=True)
raw_image = Image.open(requests.get(image_file, stream=True).raw)
inputs = processor(images=raw_image, text=prompt, return_tensors="pt")

output = model.generate(**inputs, max_new_tokens=20, do_sample=False)
print(processor.decode(output[0], skip_special_tokens=True))
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by eaidova** [comment]
@echarlaix could you please take a look? Thanks!

P.S. I'm still working on extending models coverage, but I think it makes sense to start looking on general API now

File: optimum/exporters/openvino/model_configs.py [hunk 1]
```diff
@@ -1018,3 +1035,448 @@ def patch_model_for_export(
...
+
+        if behavior == LlavaConfigBehavior.VISION_EMBEDDINGS:
+            vision_embedding = model.vision_tower
+            vision_embedding.config.output_hidden_states = True
+            vision_embedding.vision_model.config.output_hidden_states = True
+            return vision_embedding
+
+        if behavior == LlavaConfigBehavior.TEXT_EMBEDDINGS:
+            text_embedding = model.get_input_embeddings()
+            text_embedding.Llavaedding
```
Comment (eaidova) [inline optimum/exporters/openvino/model_configs.py:1321]:
```suggestion
           return text_embedding
```

File: optimum/exporters/openvino/model_configs.py [hunk 2]
```diff
@@ -1018,3 +1035,448 @@ def patch_model_for_export(
...
+
+        if behavior == LlavaConfigBehavior.VISION_EMBEDDINGS:
+            vision_embedding = model.vision_tower
+            vision_embedding.config.output_hidden_states = True
+            vision_embedding.vision_model.config.output_hidden_states = True
+            return vision_embedding
+
+        if behavior == LlavaConfigBehavior.TEXT_EMBEDDINGS:
+            text_embedding = model.get_input_embeddings()
+           return text_embedding
```
Comment (eaidova) [inline optimum/exporters/openvino/model_configs.py:1321]:
```suggestion
           text_embedding.config = model.config.text_config
           return text_embedding
```

File: optimum/exporters/openvino/model_configs.py [hunk 3]
```diff
@@ -1018,3 +1035,449 @@ def patch_model_for_export(
...
+        if behavior == LlavaConfigBehavior.VISION_EMBEDDINGS:
+            vision_embedding = model.vision_tower
+            vision_embedding.config.output_hidden_states = True
+            vision_embedding.vision_model.config.output_hidden_states = True
+            return vision_embedding
+
+        if behavior == LlavaConfigBehavior.TEXT_EMBEDDINGS:
+            text_embedding = model.get_input_embeddings()
+           text_embedding.config = model.config.text_config
+           return text_embedding
```
Comment (eaidova) [inline optimum/exporters/openvino/model_configs.py:1322]:
```suggestion
            text_embedding = model.get_input_embeddings()
            text_embedding.config = model.config.text_config
            return text_embedding
```

File: optimum/intel/openvino/modeling_visual_language.py [hunk 4]
```diff
@@ -0,0 +1,1205 @@
...
+        local_files_only: bool = False,
+        task: Optional[str] = None,
+        use_cache: bool = True,
+        trust_remote_code: bool = False,
+        load_in_8bit: Optional[bool] = None,
+        quantization_config: Union[OVWeightQuantizationConfig, Dict] = None,
+        **kwargs,
+    ):
+        """
+        Export a vanilla Transformers model into an ONNX model using `transformers.onnx.export_onnx`.
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_visual_language.py:532]: Do we really export to the ONNX? The docstrings should be revised accordingly.

**Review by AlexKoff88 — APPROVED** [review]

File: optimum/intel/openvino/modeling_visual_language.py (hunk 4 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py]: I accidentally copied that from https://github.com/huggingface/optimum-intel/blob/main/optimum/intel/openvino/modeling_base.py#L534
I see the same doc string in other model classes for openvino (seq2seq, stable diffusion support e.t.c.), @echarlaix  maybe it is time to revise them in all models too?

File: optimum/exporters/openvino/convert.py [hunk 5]
```diff
@@ -687,7 +697,7 @@ def export_from_model(
         input_shapes=input_shapes,
         device=device,
         ov_config=ov_config,
-        stateful=stateful,
+        stateful=stateful_for_parts,
```
[the remark this answers is not in the document]
Comment (eaidova) [reply optimum/exporters/openvino/convert.py]: done

File: optimum/exporters/openvino/convert.py [hunk 6]
```diff
@@ -778,3 +788,84 @@ def _add_version_info_to_model(model: Model, library_name: Optional[str] = None)
...
+    models_for_export = {}
+    stateful_parts = []
+
+    model_type = model.config.model_type.replace("_", "-")
+
+    if model_type == "internvl-chat" and preprocessors is not None:
+        model.config.img_context_token_id = preprocessors[0].convert_tokens_to_ids("<IMG_CONTEXT>")
+
+    if hasattr(model, "image_newline"):
+        model.config.image_newline = model.image_newline.tolist()
```
Comment (echarlaix) [inline optimum/exporters/openvino/convert.py:812]: why is this needed ?

File: optimum/exporters/openvino/convert.py [hunk 7]
```diff
@@ -778,3 +788,84 @@ def _add_version_info_to_model(model: Model, library_name: Optional[str] = None)
...
+    main_config = main_config_cls(
+        model.config, int_dtype=int_dtype, float_dtype=float_dtype, preprocessors=preprocessors
+    )
+    for behavior in main_config.SUPPORTED_BEHAVIORS:
+        model_id = f"{behavior}_model"
+        model_part_config = main_config.with_behavior(behavior)
+        model_part = main_config.get_model_for_behaviour(model, behavior)
+        models_for_export[model_id] = (model_part, model_part_config)
+        stateful_parts.append(stateful if getattr(model_part_config, "use_past", False) else False)
+    return main_config, models_for_export, stateful_parts
```
Comment (echarlaix) [inline optimum/exporters/openvino/convert.py:825]: `stateful_parts` can be inferred from `models_for_export` so can't we get this value from there instead of returning `stateful_parts` here and in `_get_submodels_and_export_configs` to pass it to `export_models` afterwards ?

File: optimum/exporters/openvino/model_patcher.py [hunk 8]
```diff
@@ -2652,3 +2652,53 @@ def __init__(
...
+        model.forward = model.extract_feature
+
+        super().__init__(config, model, model_kwargs)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        super().__exit__(exc_type, exc_value, traceback)
+        self._model.forward = self._model.__orig_forward
+
+
+def llava_vision_embed_forward(self, pixel_values):
```
Comment (echarlaix) [inline optimum/exporters/openvino/model_patcher.py:2674]: can you add the link to the original code + specify what has been modified so that it's later easier for maintenance ?

File: optimum/intel/utils/modeling_utils.py [hunk 9]
```diff
@@ -267,6 +267,8 @@ def _infer_library_from_model_or_model_class(
 ):
     if model.__module__.startswith("open_clip"):
         library_name = "open_clip"
+    elif model.__module__.startswith("torch"):
+        library_name = "transformers"
```
Comment (echarlaix) [inline optimum/intel/utils/modeling_utils.py:271]: why this ?

File: optimum/exporters/openvino/model_configs.py [hunk 10]
```diff
@@ -1194,3 +1210,372 @@ def patch_model_for_export(
...
+
+    def generate_dummy_inputs(self, framework: str = "pt", **kwargs) -> Dict:
+        if self._behavior == LlavaConfigBehavior.VISION_EMBEDDINGS and self._config.model_type == "pixtral":
+            kwargs["batch_size"] = 1
+        return super().generate_dummy_inputs(framework, **kwargs)
+
+
+@register_in_tasks_manager("llava-next", *["image-text-to-text"], library_name="transformers")
+class LlavaNextOpenVINOConfig(LlavaOpenVINOConfig):
+    pass
```
Comment (echarlaix) [inline optimum/exporters/openvino/model_configs.py:1429]: we could add the minimum transformers version needed with MIN_TRANSFORMERS_VERSION

File: optimum/exporters/openvino/model_configs.py [hunk 11]
```diff
@@ -1194,3 +1210,372 @@ def patch_model_for_export(
...
+            return self.__class__(
+                self._orig_config,
+                task=self.task,
+                int_dtype=self.int_dtype,
+                float_dtype=self.float_dtype,
+                behavior=behavior,
+                preprocessors=self._preprocessors,
+            )
+
+    def get_model_for_behaviour(self, model, behavior: Union[str, LlavaConfigBehavior]):
```
Comment (echarlaix) [inline optimum/exporters/openvino/model_configs.py:1560]:
```suggestion
    def get_model_for_behavior(self, model, behavior: Union[str, LlavaConfigBehavior]):
```

File: optimum/intel/openvino/modeling_visual_language.py [hunk 12]
```diff
@@ -0,0 +1,1185 @@
...
+        self.request.wait()
+        logits = self.request.get_tensor("logits").data
+        logits = torch.from_numpy(logits).to(self.device)
+        past_key_values = ((),)
+        self._past_length += inputs["inputs_embeds"].shape[1]
+
+        return CausalLMOutputWithPast(logits=logits, past_key_values=past_key_values)
+
+
+class OVModelPart:
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:169]: could we re-use part of https://github.com/huggingface/optimum-intel/blob/9cc0ee41b5e9ef114afeaeb11ba715d4165b7d74/optimum/intel/openvino/modeling_diffusion.py#L634 and inherit from it (base class could be moved in https://github.com/huggingface/optimum-intel/blob/9cc0ee41b5e9ef114afeaeb11ba715d4165b7d74/optimum/intel/openvino/modeling_base.py)

File: optimum/intel/openvino/modeling_visual_language.py [hunk 13]
```diff
@@ -0,0 +1,1185 @@
...
+        if self.text_emb_request is None:
+            logger.info(f"Compiling the Text embeddings model to {self._device} ...")
+            self.text_emb_request = core.compile_model(self.text_emb_model, self._device, self.ov_config)
+
+    def to(self, device: str):
+        if isinstance(device, str):
+            self._device = device.upper()
+            self.clear_requests()
+
+        return self
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:68]: could make sense to modify `OVModel.to` to not have to overwrite it https://github.com/huggingface/optimum-intel/blob/9cc0ee41b5e9ef114afeaeb11ba715d4165b7d74/optimum/intel/openvino/modeling.py#L129 (just adding a `clear_requests` method), what do you think ?

**Review by echarlaix — COMMENTED** [review]
Looking good, thanks a lot @eaidova

File: optimum/exporters/openvino/convert.py (hunk 6 above)
Comment (eaidova) [reply optimum/exporters/openvino/convert.py:812]: this is tranable parameter filled by random values during model weights init, we can not capture this during model export as it is not a part of submodels inference forward pass, that is why for transferring it from pt to ov I need to use config for that. If you have better suggestion, I can try to implement this

File: optimum/exporters/openvino/model_configs.py (hunk 10 above)
Comment (eaidova) [reply optimum/exporters/openvino/model_configs.py:1429]: I think it can be complicated e.g. pixtral model is also has model_type llava, but it is required 4.45 to create such model, while usual llava added in 4.39 if I'm not wrong

File: optimum/exporters/openvino/convert.py (hunk 7 above)
Comment (eaidova) [reply optimum/exporters/openvino/convert.py:825]: I think it maybe nontrivial and more complex if we move that on other level, where some info about model type can be already lost. e.g. seq2seq models also split on parts and have decoder with past that currently should remain stateless until ov runtime will be ready for this IR format https://github.com/huggingface/optimum-intel/pull/779 (yes, we still working on that and I belive it will happens soon). also with exploring more models, it maybe changed too, e.g. there is blip2-t5-xl model that uses encoder_decoder architecture for language model with the same working approach that also will require some changes in this logic in future

File: optimum/intel/utils/modeling_utils.py (hunk 9 above)
Comment (eaidova) [reply optimum/intel/utils/modeling_utils.py:271]: model.get_input_embeddings() returns instance of torch.nn.Embeddings class, if it is possible skip this check for such models, I think we can avoid this workaround

File: optimum/intel/openvino/modeling_visual_language.py (hunk 12 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:169]: not sure about inheritage, becuase it adds some extra functional (e.g. _compile vs compile, model export as completely independent models), but agree for moving

File: optimum/exporters/openvino/model_configs.py (hunk 11 above)
Comment (eaidova) [reply optimum/exporters/openvino/model_configs.py]: done

File: optimum/intel/openvino/modeling_visual_language.py (hunk 12 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py]: done

File: optimum/intel/openvino/modeling_visual_language.py (hunk 13 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:67]: done

File: optimum/exporters/openvino/stateful.py [hunk 14]
```diff
@@ -213,14 +224,22 @@ def patch_stateful(config: PretrainedConfig, ov_model: ov.Model):
     if not key_value_input_names or not key_value_output_names:
         return
 
+    main_input_name = "input_ids" if model_has_input_output_name(ov_model, "input_ids") else "inputs_embeds"
```
[the remark this answers is not in the document]
Comment (eaidova) [reply optimum/exporters/openvino/stateful.py]: done

File: optimum/intel/utils/modeling_utils.py (hunk 9 above)
Comment (eaidova) [reply optimum/intel/utils/modeling_utils.py]: added opportunity to propagate library_name from whole model to subcomponets

File: optimum/exporters/openvino/convert.py (hunk 7 above)
Comment (eaidova) [reply optimum/exporters/openvino/convert.py:834]: done

File: optimum/exporters/openvino/model_patcher.py (hunk 8 above)
Comment (eaidova) [reply optimum/exporters/openvino/model_patcher.py:2674]: done

File: optimum/exporters/openvino/model_configs.py (hunk 10 above)
Comment (eaidova) [reply optimum/exporters/openvino/model_configs.py]: done

File: optimum/exporters/openvino/convert.py (hunk 7 above)
Comment (echarlaix) [reply optimum/exporters/openvino/convert.py:834]: Makes sense, let's keep it that way then!

> https://github.com/huggingface/optimum-intel/pull/779 (yes, we still working on that and I belive it will happens soon)

that's very cool !

File: optimum/intel/openvino/modeling_visual_language.py [hunk 15]
```diff
@@ -0,0 +1,1192 @@
...
+                Whether or not to only look at local files (i.e., do not try to download the model).
+        """
+        if use_auth_token is not None:
+            warnings.warn(
+                "The `use_auth_token` argument is deprecated and will be removed soon. Please use the `token` argument instead.",
+                FutureWarning,
+            )
+            if token is not None:
+                raise ValueError("You cannot use both `use_auth_token` and `token` arguments at the same time.")
+            token = use_auth_token
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:380]: shouldn't be needed I think as already check in https://github.com/huggingface/optimum-intel/blob/6a8718b76a5fc5e6796483746eded64e4007e94d/optimum/intel/openvino/modeling_base.py#L394

```suggestion
```

File: optimum/intel/openvino/modeling_visual_language.py [hunk 16]
```diff
@@ -0,0 +1,1192 @@
...
+    ):
+        if use_auth_token is not None:
+            warnings.warn(
+                "The `use_auth_token` argument is deprecated and will be removed soon. Please use the `token` argument instead.",
+                FutureWarning,
+            )
+            if token is not None:
+                raise ValueError("You cannot use both `use_auth_token` and `token` arguments at the same time.")
+            token = use_auth_token
+
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:541]:
```suggestion
```

File: optimum/intel/openvino/modeling_visual_language.py [hunk 17]
```diff
@@ -0,0 +1,1192 @@
...
+            text_embeddings=text_embeddings,
+            vision_embeddings=vision_embeddings,
+            config=config,
+            model_save_dir=model_save_dir,
+            quantization_config=quantization_config,
+            **kwargs,
+        )
+
+    @classmethod
+    def _from_transformers(
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:515]: what is the main difference with https://github.com/huggingface/optimum-intel/blob/6a8718b76a5fc5e6796483746eded64e4007e94d/optimum/intel/openvino/modeling_decoder.py#L262 ? If only the "with-past" suffix maybe we can change it there instead wdyt ?

File: optimum/intel/openvino/modeling_visual_language.py [hunk 18]
```diff
@@ -0,0 +1,1192 @@
...
+                    kwargs.get("ov_config"),
+                    model_save_dir,
+                )
+                for part in model_cls.additional_parts:
+                    kwargs[part] = model_cls._compile_model(
+                        file_names[part],
+                        kwargs.get("device", "CPU"),
+                        kwargs.get("ov_config"),
+                        model_save_dir,
+                    )
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:490]: could the code be simplified with something like 

```suggestion
        model_file_names = {
            "language_model": "openvino_language_model.xml",
            "text_embeddings":  "openvino_text_embeddings_model.xml",
            "vision_embeddings":  "openvino_vision_embeddings_model.xml",
        }
    
        for part in model_cls.additional_parts:
            model_file_names[part] = f"openvino_{part}_model.xml"

        model_cls = MODEL_TYPE_TO_CLS_MAPPING[config.model_type]

        quantization_config = model_cls._prepare_weight_quantization_config(quantization_config, load_in_8bit)
        compile_only = kwargs.get("compile_only", False)

        if os.path.isdir(model_id):
            model_save_dir = Path(model_id)
            file_names = {k:os.path.join(model_id, model_file_names[k]) for k in model_file_names}
        else:
            file_names = model_file_names.copy()
            for name, file_name in model_file_names.items():
                model_cache_path = hf_hub_download(
                    repo_id=model_id,
                    filename=file_name,
                    token=token,
                    revision=revision,
                    cache_dir=cache_dir,
                    force_download=force_download,
                    local_files_only=local_files_only,
                )
                file_names[name] = model_cache_path

            model_save_dir = Path(model_cache_path).parent

        if not compile_only:
            language_model = model_cls.load_model(file_names["language_model"], quantization_config)
            text_embeddings = model_cls.load_model(file_names["text_embeddings"], quantization_config)
            vision_embeddings = model_cls.load_model(file_names["vision_emnbeddings"], quantization_config)
            for part in model_cls.additional_parts:
                kwargs[part] = model_cls.load_model(file_names[part], quantization_config)
        else:
            language_model = model_cls._compile_model(
                file_names["language_model"],
                kwargs.get("device", "CPU"),
                kwargs.get("ov_config"),
                model_save_dir,
            )
            text_embeddings = model_cls._compile_model(
                file_names["text_embeddings"],
                kwargs.get("device", "CPU"),
                kwargs.get("ov_config"),
                model_save_dir,
            )
            vision_embeddings = model_cls._compile_model(
                file_names["vision_embeddings"],
                kwargs.get("device", "CPU"),
                kwargs.get("ov_config"),
                model_save_dir,
            )
            for part in model_cls.additional_parts:
                kwargs[part] = model_cls._compile_model(
                    file_names[part],
                    kwargs.get("device", "CPU"),
                    kwargs.get("ov_config"),
                    model_save_dir,
                )
```

File: optimum/intel/openvino/modeling_visual_language.py [hunk 19]
```diff
@@ -0,0 +1,1192 @@
...
+                        patch_feats.append(self.vision_embeddings(np.expand_dims(patch_value, 0)).last_hidden_state)
+                    patch_feats = np.concatenate(patch_feats, axis=1)
+                else:
+                    patch_feats = self.vision_embeddings(patch).last_hidden_state
+                image_features.append(patch_feats)
+            image_features = np.concatenate(image_features, 0)
+
+        return image_features
+
+    def merge_vision_text_embeddings(
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_visual_language.py:750]: could you add a link to the original code for maintenance ? https://github.com/huggingface/transformers/blob/d7950bff82b18c823193d17d72188c5e46d06c83/src/transformers/models/llava/modeling_llava.py#L297C9-L297C45

**Review by echarlaix — APPROVED** [review]
Huge work, thanks @eaidova, left couple of minor comments

File: optimum/intel/openvino/modeling_visual_language.py (hunk 17 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:515]: do you mean I should inherit from OVModelForCausalLM to get this behaviour? I think it may lead to issues with quantization as it may requires slightly different approach for doing that (difference in input data collection, different approaches for vision embeddings and text model and so on)

File: optimum/intel/openvino/modeling_visual_language.py [hunk 20]
```diff
@@ -0,0 +1,1192 @@
...
+        if self._compile_only:
+            raise ValueError(
+                "`to()` is not supported with `compile_only` mode, please intialize model without this option"
+            )
+
+        if isinstance(device, str):
+            self._device = device.upper()
+            self.clear_requests()
+
+        return self
```
[the remark this answers is not in the document]
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py]: removed

File: optimum/intel/openvino/modeling_visual_language.py (hunk 15 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:365]: removed

File: optimum/intel/openvino/modeling_visual_language.py (hunk 16 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py]: removed

File: optimum/intel/openvino/modeling_visual_language.py (hunk 18 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py]: simplified

File: optimum/intel/openvino/modeling_visual_language.py (hunk 19 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:690]: added

File: optimum/intel/openvino/modeling_visual_language.py (hunk 17 above)
Comment (echarlaix) [reply optimum/intel/openvino/modeling_visual_language.py:459]: makes sense, let's keep it that way then!

**Comment by echarlaix** [comment]
> P.S. I'm still working on extending models coverage, but I think it makes sense to start looking on general API now

Very nice, let's merge this PR and we can always extend support in following PRs, does it work for you @eaidova ?

**Comment by eaidova** [comment]
> > P.S. I'm still working on extending models coverage, but I think it makes sense to start looking on general API now
> 
> Very nice, let's merge this PR and we can always extend support in following PRs, does it work for you @eaidova ?

@echarlaix , yes, I agree, could you please merge?
